### PR TITLE
Closes #5

### LINKED ISSUE #5: Remove dylint dependency and instead provide our own CLI interface
Dylint is great, but makes it quite difficult to use in CI due to the need for dynamic loading. Instead we'd like to cannibalize clippy's infrastructure to provide a single binary.

### PR DESCRIPTION
**Pull request by TheCodingWombat**
missing-security-doc lint working.

Test using `cargo test`

### CODE REVIEW

**Comment by TheCodingWombat** [comment]
Closes #5

File: .editorconfig [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+indent_style = space
+indent_size = 4
+
+[*.md]
+# double whitespace at end of line
+# denotes a line break in Markdown
+trim_trailing_whitespace = false
+
+[*.yml]
+indent_size = 2
```
Comment (KaiserKarel) [inline .editorconfig]: Prefer not to check it in, and instead use linters to guarantee formatting in CI.

File: .gitattributes [hunk 2]
```diff
@@ -0,0 +1,3 @@
+* text=auto eol=lf
+*.rs text eol=lf whitespace=tab-in-indent,trailing-space,tabwidth=4
+*.fixed linguist-language=Rust
```
Comment (KaiserKarel) [inline .gitattributes]: Should this be checked in?

File: Cargo.toml [hunk 3]
```diff
@@ -1,25 +1,69 @@
 [package]
 name = "substrace"
-version = "0.1.0"
-authors = ["authors go here"]
-description = "decription goes here"
-edition = "2018"
-publish = false
+version = "0.0.0"
+description = "A bunch of helpful lints to avoid common pitfalls in Rust"
+repository = "https://github.com/rust-lang/rust-clippy"
```
Comment (KaiserKarel) [inline Cargo.toml]: These fields need to be updated

File: Cargo.toml [hunk 4]
```diff
@@ -1,25 +1,69 @@
...
+readme = "README.md"
+license = "MIT OR Apache-2.0"
+keywords = ["substrace", "lint", "plugin"]
+categories = ["development-tools", "development-tools::cargo-plugins"]
+build = "build.rs"
+edition = "2021"
+publish = true
```
Comment (KaiserKarel) [inline Cargo.toml]: Let's set publish to false until we can publish through CI

**Review by KaiserKarel — CHANGES_REQUESTED** [review]
Seems this PR still has a lot of files from Clippy, which should be removed.

File: README.md [hunk 5]
```diff
@@ -1,22 +1,255 @@
-# Substrace: Experimental Static Analysis for Substrate Projects
+# Clippy
 
-[Substrace](https://github.com/kaiserkarel/substrace) is a tool for linting substrate projects for specific gotchas and other errors.
+[![Clippy Test](https://github.com/rust-lang/rust-clippy/workflows/Clippy%20Test/badge.svg?branch=auto&event=push)](https://github.com/rust-lang/rust-clippy/actions?query=workflow%3A%22Clippy+Test%22+event%3Apush+branch%3Aauto)
```
Comment (KaiserKarel) [inline README.md]: Let's use the old README for now, and remove dylint reference

File: book/README.md [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# Clippy Book
+
+This is the source for the Clippy Book. See the
+[book](src/development/infrastructure/book.md) for more information.
```
Comment (KaiserKarel) [inline book/README.md]: Entire book can be removed

**Review by TheCodingWombat — COMMENTED** [review]
I removed all unused clippy stuff

File: substrace_lints/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,274 @@
...
+#[macro_use]
+extern crate substrace_utils;
+
+use substrace_utils::parse_msrv;
+use rustc_data_structures::fx::FxHashSet;
+use rustc_lint::LintId;
+use rustc_semver::RustcVersion;
+use rustc_session::Session;
+
+/// Macro used to declare a Clippy lint.
```
Comment (KaiserKarel) [inline substrace_lints/src/lib.rs]:
```suggestion
/// Macro used to declare a Substrace lint.
```

Comment (KaiserKarel) [inline substrace_lints/src/lib.rs]:
```suggestion
/// Macro used to declare a Substrace lint.
```

File: src/driver.rs [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+
+        let clippy_enabled = !cap_lints_allow && (!no_deps || in_primary_package);
+        if clippy_enabled {
+            args.extend(clippy_args);
+            rustc_driver::RunCompiler::new(&args, &mut ClippyCallbacks { clippy_args_var }).run()
+        } else {
+            rustc_driver::RunCompiler::new(&args, &mut RustcCallbacks { clippy_args_var }).run()
+        }
+    }))
+}
```
Comment (KaiserKarel) [inline src/driver.rs]: Let's refactor all references to clippy to Substrace here for consistency.

File: src/main.rs [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+#![warn(rust_2018_idioms, unused_lifetimes)]
+
+use rustc_tools_util::VersionInfo;
+use std::env;
+use std::path::PathBuf;
+use std::process::{self, Command};
+
+mod docs;
+
+const CARGO_CLIPPY_HELP: &str = r#"Substrace: Checks a package to catch common mistakes and improve your Rust code.
```
Comment (KaiserKarel) [inline src/main.rs]: Same here, let's remove all references to clippy and rename to substrace

File: substrace_lints/README.md [hunk 10]
```diff
@@ -0,0 +1 @@
+This crate contains Clippy lints. For the main crate, check [GitHub](https://github.com/rust-lang/rust-clippy).
```
Comment (KaiserKarel) [inline substrace_lints/README.md]: Fix link and name

File: substrace_lints/Cargo.toml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+unicode-script = { version = "0.5", default-features = false }
+semver = "1.0"
+rustc-semver = "1.1"
+# NOTE: cargo requires serde feat in its url dep
+# see <https://github.com/rust-lang/rust/pull/63587#issuecomment-522343864>
+url = { version = "2.2", features = ["serde"] }
+
+[features]
+deny-warnings = ["substrace_utils/deny-warnings"]
+# build clippy with internal lints enabled, off by default
```
Comment (KaiserKarel) [inline substrace_lints/Cargo.toml]: fix comment

File: substrace_lints/src/substrace_lints/auxiliary/paths.rs [hunk 12]
```diff
@@ -1,20 +1,19 @@
...
     for typ in generics.args {
         if let hir::GenericArg::Type(ty) = typ {
             if let hir::TyKind::Path(hir::QPath::Resolved(_, path)) = &ty.kind {
                 if let hir::def::Res::Def(hir::def::DefKind::Struct, id) = path.res {
-                    return match_def_path(cx, id, &TWOX64CONCTAT)
+                    if match_def_path(cx, id, &TWOX64CONCTAT)
                         || match_def_path(cx, id, &IDENTITY)
                         || match_def_path(cx, id, &TWOX128)
-                        || match_def_path(cx, id, &TWOX256);
+                        || match_def_path(cx, id, &TWOX256) {
```
Comment (KaiserKarel) [inline substrace_lints/src/substrace_lints/auxiliary/paths.rs:14]: Identity is also a dangerous hasher

File: substrace_lints/src/utils/author.rs [hunk 13]
```diff
@@ -0,0 +1,750 @@
...
+    ///     if x == 42 {
+    ///         // but ignore everything from here on
+    ///         #![clippy::author = "ignore"]
+    ///     }
+    ///     ()
+    /// }
+    /// ```
+    ///
+    /// Running `TESTNAME=ui/my_lint cargo uitest` will produce
+    /// a `./tests/ui/new_lint.stdout` file with the generated code:
```
Comment (KaiserKarel) [inline substrace_lints/src/utils/author.rs]: Fix references to clippy.

File: substrace_utils/src/diagnostics.rs [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+                "for further information visit https://rust-lang.github.io/rust-clippy/{}/index.html#{}",
+                &option_env!("RUST_RELEASE_NUM").map_or("master".to_string(), |n| {
+                    // extract just major + minor version and ignore patch versions
+                    format!("rust-{}", n.rsplit_once('.').unwrap().1)
+                }),
+                lint
+            ));
+        }
+    }
+}
```
Comment (KaiserKarel) [inline substrace_utils/src/diagnostics.rs]: Fix references to clippy

File: substrace_utils/src/paths.rs [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+pub const COW: [&str; 3] = ["alloc", "borrow", "Cow"];
+pub const CORE_ITER_COLLECT: [&str; 6] = ["core", "iter", "traits", "iterator", "Iterator", "collect"];
+pub const CORE_ITER_CLONED: [&str; 6] = ["core", "iter", "traits", "iterator", "Iterator", "cloned"];
+pub const CORE_ITER_COPIED: [&str; 6] = ["core", "iter", "traits", "iterator", "Iterator", "copied"];
+pub const CORE_ITER_FILTER: [&str; 6] = ["core", "iter", "traits", "iterator", "Iterator", "filter"];
+pub const CORE_ITER_INTO_ITER: [&str; 6] = ["core", "iter", "traits", "collect", "IntoIterator", "into_iter"];
+pub const CSTRING_AS_C_STR: [&str; 5] = ["alloc", "ffi", "c_str", "CString", "as_c_str"];
+pub const DEFAULT_TRAIT_METHOD: [&str; 4] = ["core", "default", "Default", "default"];
+pub const DEREF_MUT_TRAIT_METHOD: [&str; 5] = ["core", "ops", "deref", "DerefMut", "deref_mut"];
+/// Preferably use the diagnostic item `sym::deref_method` where possible
```
Comment (KaiserKarel) [inline substrace_utils/src/paths.rs]: Are all of these paths required for our lints?

File: tests/compile-test.rs [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+    "if_chain",
+    "itertools",
+    "quote",
+    "regex",
+    "serde",
+    "serde_derive",
+    "syn",
+    "tokio",
+    "parking_lot",
+    "rustc_semver",
```
Comment (KaiserKarel) [inline tests/compile-test.rs]: Let's not test these lints/remove them?

**Review by KaiserKarel — CHANGES_REQUESTED** [review]
After addressing comments, this can be merged :)

File: tests/compile-test.rs (hunk 16 above)
Comment (TheCodingWombat) [reply tests/compile-test.rs]: Indeed not necessary. Removed.

File: substrace_utils/src/paths.rs (hunk 15 above)
Comment (TheCodingWombat) [reply substrace_utils/src/paths.rs]: Most not necessary. Removed all that are not necessary for compilation.

**Review by KaiserKarel — APPROVED** [review]

**Pull request merged**